tools/promql-generator/promql-generator.go

```diff
@@ -0,0 +1,354 @@
+// Note - the matcher will only find a label set used in the last 2 hours
+func labelSetsFromPrometheus(address string, matcher string) ([]model.LabelSet, error) {
+	client, err := api.NewClient(api.Config{
+		Address: address,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "create Prometheus client")
+	}
+	promAPI := v1.NewAPI(client)
+	ctx := context.Background()
+	now := time.Now()
+
+	labels, _, err := promAPI.Series(
+		ctx,
+		[]string{matcher},
+		now.Add(-2*time.Hour),
+		now,
+	)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "get label sets")
+	}
+
+	if len(labels) == 0 {
+		return nil, errors.New("no matching label sets found")
+	}
+
+	return labels, nil
+}
+
+// labelSetsFromFile constructs a []model.LabelSet from a json file defining the label sets
+func labelSetsFromFile(seriesFile string) ([]model.LabelSet, error) {
+	data, err := os.ReadFile(seriesFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var result []map[string]string
+	if err := json.Unmarshal(data, &result); err != nil {
+		return nil, err
+	}
+
+	if len(result) == 0 {
+		return nil, errors.New("no label sets found")
+	}
+
+	labels := make([]model.LabelSet, len(result))
+
+	for i, item := range result {
+		labelSet := make(model.LabelSet)
+		for k, v := range item {
+			labelSet[model.LabelName(k)] = model.LabelValue(v)
+		}
+		labels[i] = labelSet
+	}
+
+	return labels, nil
+}
+
+// generatePromQL generate the random PromQL queries for the given label sets and writes to stdout
+func generatePromQL(count int, labelSets []model.LabelSet) error {
+
+	now := time.Now()
+	rnd := rand.New(rand.NewSource(now.Unix()))
+	opts := []promqlsmith.Option{
+		promqlsmith.WithEnableOffset(true),
+		promqlsmith.WithEnableAtModifier(true),
+		promqlsmith.WithEnabledFunctions(availableFunctions()),
+		promqlsmith.WithEnabledBinOps(enabledBinops),
+		promqlsmith.WithEnableVectorMatching(true),
+	}
+	ps := promqlsmith.New(rnd, modelLabelSetToLabels(labelSets), opts...)
+
+	for range count {
+
+		expr := ps.WalkInstantQuery()
+		query := expr.Pretty(0)
+
+		fmt.Printf("%s\n", strings.ReplaceAll(query, "\n", ""))
+
+	}
+	return nil
+}
+
+// modelLabelSetToLabels converts []model.LabelSet to []labels.Labels
+func modelLabelSetToLabels(labelSets []model.LabelSet) []labels.Labels {
+	out := make([]labels.Labels, len(labelSets))
+	bufLabels := labels.EmptyLabels()
+	builder := labels.NewBuilder(bufLabels)
+	for i, lbls := range labelSets {
+		for k, v := range lbls {
+			builder.Set(string(k), string(v))
+		}
+		out[i] = builder.Labels()
+		builder.Reset(bufLabels)
+	}
+	return out
+}
+
+// getAvailableFunctions returns a []*parser.Function excluding functions listed in the unsupportedFunctions map
+func availableFunctions() []*parser.Function {
+	var results []*parser.Function
+	for _, f := range parser.Functions {
+		if f.Variadic != 0 {
+			continue
+		}
+		if slices.Contains(f.ArgTypes, parser.ValueTypeString) {
+			continue
+		}
```

Comment:
Might be worth adding a comment explaining why we skip functions in these two cases.